### PR TITLE
Implement LINKED-DTC-DOPs

### PR DESCRIPTION
**Pull request by andlaus**
This turned out to be a bit more involved than anticipated: In contrast to what the term "linked" implies, linked DTC DOPs are not just informative, but an inheritance mechanism for DTCs...

Andreas Lauser &lt;andreas.lauser@mercedes-benz.com&gt;, on behalf of [MBition GmbH](https://mbition.io/).
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

### CODE REVIEW

File: odxtools/dtcdop.py [hunk 1]
```diff
@@ -201,14 +248,52 @@ def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
...
-        self._linked_dtc_dops = NamedItemList(linked_dtc_dops)
+        for linked_dtc_dop in self.linked_dtc_dops_raw:
+            linked_dtc_dop._resolve_odxlinks(odxlinks)
 
     def _resolve_snrefs(self, context: SnRefContext) -> None:
+        # hack to avoid initializing the DtcDop object multiple
+        # times. This is required, because the linked DTC DOP feature
+        # requires the "parent" DTC DOPs to be fully initialized but
+        # the standard does not define a formal ordering of DTC DOPs.
+        if hasattr(self, "_linked_dtc_dops"):
```
Comment (kayoub5) [inline odxtools/dtcdop.py:259]: hasattr have performance impact and mypy can't check it, better to use a bool attribute set it to False in init, and True here

File: odxtools/dtcdop.py [hunk 2]
```diff
@@ -201,14 +248,52 @@ def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
...
 
         for dtc_proxy in self.dtcs_raw:
             if isinstance(dtc_proxy, DiagnosticTroubleCode):
                 dtc_proxy._resolve_snrefs(context)
+
+        for linked_dtc_dop in self.linked_dtc_dops_raw:
+            linked_dtc_dop._resolve_snrefs(context)
+
+        # add the inherited DTCs from linked DTC DOPs. Note that this
+        # requires that there are no cycles in the "link-hierarchy"
```
Comment (kayoub5) [inline odxtools/dtcdop.py:274]: required by standard, or current logic?

File: odxtools/dtcdop.py [hunk 3]
```diff
@@ -201,14 +248,52 @@ def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
...
+        dtc_short_names = {dtc.short_name for dtc in self._dtcs}
+        for linked_dtc_dop in self.linked_dtc_dops_raw:
+            # hack to make sure that the dtcs of the linked DTC DOP
+            # objects have been determined.
+            if not hasattr(linked_dtc_dop.dtc_dop, "_linked_dtc_dops"):
```
Comment (kayoub5) [inline odxtools/dtcdop.py:279]: This check really needed? `_resolve_snrefs` already perform this check

File: odxtools/dtcdop.py [hunk 4]
```diff
@@ -30,7 +76,7 @@ class DtcDop(DopBase):
     physical_type: PhysicalType
     compu_method: CompuMethod
     dtcs_raw: List[Union[DiagnosticTroubleCode, OdxLinkRef]]
-    linked_dtc_dop_refs: List[OdxLinkRef]
+    linked_dtc_dops_raw: List[LinkedDtcDop]
```
Comment (kayoub5) [inline odxtools/dtcdop.py:79]: The linked dtcs don't affect serialization logic?

File: odxtools/dtcdop.py (hunk 1 above)
Comment (andlaus) [reply odxtools/dtcdop.py:259]: IMO the performance impact is not relevevant here: this function is only called once per `DtcDop` when loading a database, IOW typically a few dozen times. I guess I'll implement the solution which you've proposed tomorrow in order to see if I like it better than the current hack. stay tuned...

File: odxtools/dtcdop.py (hunk 2 above)
Comment (andlaus) [reply odxtools/dtcdop.py:274]: If I didn't miss something, the standard does not mention anything about "linked-dop-ref cycles".

That said, although this is not explicitly mentioned in the text, I got the impression that the standard assumes that all linked DTC DOPs are defined in one of the parent diag layers of the layer which contains the DTC-DOP in question...

Comment (andlaus) [reply odxtools/dtcdop.py:274]: (non-cyclic link graphs are required to get well-defined results, or at least the standard would need to specify how to handle the corner cases that emerge from them.)

File: odxtools/dtcdop.py (hunk 4 above)
Comment (andlaus) [reply odxtools/dtcdop.py:79]: they do, that's why `printDOP.xml.jinja2` is modified by this commit...

Comment (kayoub5) [reply odxtools/dtcdop.py:79]: I am asking if they affect encoding / decoding messages

Comment (andlaus) [reply odxtools/dtcdop.py:79]: they do not directly affect en-/decoding: just make additional DTCs appear for value parameters that use a DTC-DOP, so additional DTCs can be decoded.

File: odxtools/dtcdop.py (hunk 2 above)
Comment (andlaus) [reply odxtools/dtcdop.py:279]: okay, [changed]( https://github.com/mercedes-benz/odxtools/pull/360/commits/bf81d92c8830238ac0cc19d2d3d388a9e0669848): I played around with the idea of a separate `_is_initialized` member and I think that the result is not worse than the `hasattr()` route (IMO it is not better either, but ...)

**Review by kayoub5 — APPROVED** [review]

**Pull request merged**